src/main/com/smithsmodding/armory/common/factories/ArmorFactory.java

```diff
@@ -35,69 +37,141 @@ public static ArmorFactory getInstance() {
 
         return combinedArmor;
     }
 
     /**
      * Function used to create a new armor ItemStack
      *
-     * @param armor           The base armor used to create the new armor.
-     * @param addons          The new addons stored in a ArrayList
-     * @param totalDurability
-     * @param coreMaterial
-     * @param data            Extra data for your factories  @return A new ItemStack with full getDurability
+     * @param armor The base armor used to create the new armor.
+     * @param coreMaterial The core material of the armor.
+     * @param addons    The new addons stored in a ArrayList
+     * @return A new ItemStack with full durability.
+     * @throws IllegalArgumentException When the Addons are not compatible, or when the durability is smaller or equal to 0
      */
     @Nonnull
     @Override
-    public ItemStack buildNewMLAArmor(IMultiComponentArmor armor, ArrayList<IMultiComponentArmorExtensionInformation> addons, Integer totalDurability, ICoreArmorMaterial coreMaterial, Object... data) throws IllegalArgumentException {
+    public ItemStack buildNewMLAArmor(IMultiComponentArmor armor, ICoreArmorMaterial coreMaterial, List<IMultiComponentArmorExtensionInformation> addons)
+      throws IllegalArgumentException
+    {
         if (!validateNewAgainstNewAddons(addons))
             throw new IllegalArgumentException("ADDONS not compatible");
 
+        final int totalDurability = ArmorHelper.getModifyableCapabilityValue(armor, coreMaterial, addons, ModCapabilities.MOD_ARMOR_DURABILITY_CAPABILITY).intValue();
+        if (totalDurability <= 0)
+        {
+            throw new IllegalArgumentException(String.format("Combination of armor, material and extensions results in a <= 0 durability: %s,%s, %s", armor, coreMaterial, addons));
+        }
+
         ItemStack armorStack =  new ItemStack(armor.getItem(), 1);
         IMultiComponentArmorCapability capability = new IMultiComponentArmorCapability.Impl()
                 .setArmorType(armor)
                 .setMaterial(coreMaterial)
                 .setInstalledExtensions(addons)
-                .setMaximalDurability(totalDurability)
                 .setCurrentDurability(totalDurability);
 
         armorStack.getCapability(SmithsCoreCapabilityDispatcher.INSTANCE_CAPABILITY, null).getDispatcher().registerCapability(ModCapabilities.MOD_MULTICOMPONENTARMOR_CAPABILITY, capability);
 
         return armorStack;
     }
 
+    /**
+     * Gets the custom name of the Armor.
+     *
+     * @param stack The stack to get the custom name from.
+     * @return THe custom name or null if not set.
+     */
     @Nullable
     @Override
-    public String getArmorGivenName(ItemStack pStack) {
-        return null;
+    public String getArmorCustomName(@Nonnull final ItemStack stack)
+    {
+        if (!stack.hasTagCompound())
+        {
+            return null;
+        }
+
+        if (!stack.getTagCompound().hasKey("customname"))
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "customname" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1192)